C/mcal/SPI/SPI_program.c

```diff
@@ -0,0 +1,208 @@
+    SET_BIT(SSPCON1_REG, SSPCON1_CKP);
+    CLR_BIT(SSPSTAT_REG, SSPSTAT_CKE);
+    #endif
+
+    //Sample mode
+    #if SAMPLE_MODE == END
+    SET_BIT(SSPSTAT_REG, SSPSTAT_SMP);
+    #elif SAMPLE_MODE == MIDDLE
+    CLR_BIT(SSPSTAT_REG, SSPSTAT_SMP);
+    #endif
+
+    //Configure The I/O Pins For SPI Master Mode
+    CLR_BIT(TRISC_REG, TRISC_TRISC3);//SCK
+    SET_BIT(TRISC_REG, TRISC_TRISC4);//SDI
+    CLR_BIT(TRISC_REG, TRISC_TRISC5);//SDO
+    #endif
+}
+void SPI_SLAVE_Init()
+{
+    #if MCU_TYPE == _AVR
+    CLR_BIT(SPCR, SPCR_MSTR);
+    #if SPI_DATA_ORDER == MSB_FIRST
+        CLR_BIT(SPCR, SPCR_DORD);
+    #elif SPI_DATA_ORDER == LSB_FIRST
+        SET_BIT(SPCR, SPCR_DORD);
+    #endif
+     #if SPI_CLK_POLARITY == IDLE_LOW
+         CLR_BIT(SPCR, SPCR_CPOL);
+    #elif SPI_CLK_POLARITY == IDLE_HIGH
+         SET_BIT(SPCR, SPCR_CPOL);
+    #endif
+     #if SPI_CLK_PHASE == SAMPLE_FIRST
+         CLR_BIT(SPCR, SPCR_CPHA);
+    #elif SPI_CLK_PHASE == SETUP_FIRST
+          SET_BIT(SPCR, SPCR_CPHA);
+    #endif
+     //SPI ENABLE
+    SET_BIT(SPCR, SPCR_SPE);
+    //INTERRUBT
+     #if SPI_INTERRUBT== INTERRUBT_ENABLE
+        SET_BIT(SPCR, SPCR_SPIE);
+    #elif SPI_INTERRUBT== INTERRUBT_DISABLE
+        CLR_BIT(SPCR, SPCR_SPIE);
+    #endif
+    #elif MCU_TYPE == _PIC
+    //SS pin control
+    #if SPI_SS_MODE == ENABLE
+    CLR_BIT(SSPCON1_REG, SSPCON1_SSMP0);
+    CLR_BIT(SSPCON1_REG, SSPCON1_SSMP1);
+    SET_BIT(SSPCON1_REG, SSPCON1_SSMP2);
+    CLR_BIT(SSPCON1_REG, SSPCON1_SSMP3);
+    #elif SPI_SS_MODE == DISABLE
+    SET_BIT(SSPCON1_REG, SSPCON1_SSMP0);
+    CLR_BIT(SSPCON1_REG, SSPCON1_SSMP1);
+    SET_BIT(SSPCON1_REG, SSPCON1_SSMP2);
+    CLR_BIT(SSPCON1_REG, SSPCON1_SSMP3);
+    #endif
+
+    //Synchronous Serial Port Enable bit
+    SET_BIT(SSPCON1_REG, SSPCON1_SSPEN);
+
+    //SPI BUS MODES
+    #if SPI_MODE == MODE_1
+    CLR_BIT(SSPCON1_REG, SSPCON1_CKP);
+    SET_BIT(SSPSTAT_REG, SSPSTAT_CKE);
+    #elif SPI_MODE == MODE_2
+    CLR_BIT(SSPCON1_REG, SSPCON1_CKP);
+    CLR_BIT(SSPSTAT_REG, SSPSTAT_CKE);
+    #elif SPI_MODE == MODE_3
+    SET_BIT(SSPCON1_REG, SSPCON1_CKP);
+    SET_BIT(SSPSTAT_REG, SSPSTAT_CKE);
+    #elif SPI_MODE == MODE_4
+    SET_BIT(SSPCON1_REG, SSPCON1_CKP);
+    CLR_BIT(SSPSTAT_REG, SSPSTAT_CKE);
+    #endif
+
+    //Sample mode
+    #if SAMPLE_MODE == END
+    SET_BIT(SSPSTAT_REG, SSPSTAT_SMP);
+    #elif SAMPLE_MODE == MIDDLE
+    CLR_BIT(SSPSTAT_REG, SSPSTAT_SMP);
+    #endif
+
+    //Configure The I/O Pins For SPI SLAVE Mode
+    SET_BIT(TRISC_REG, TRISC_TRISC3);//SCK
+    SET_BIT(TRISC_REG, TRISC_TRISC4);//SDI
+    CLR_BIT(TRISC_REG, TRISC_TRISC5);//SDO
+
+    //SS -> INPUT
+    SET_BIT(TRISA_REG, TRISA_TRISA5);
+    #endif
+}
+void SPI_Write(uint8_t data)
+{
+    #if MCU_TYPE == _AVR
+    SPDR=data;
+    while (GET_BIT(SPSR, SPSR_SPIF)==0)
+    {
+
+    }
+    #elif MCU_TYPE == _PIC
+    while (GET_BIT(SSPSTAT_REG, SSPSTAT_BF) == 0)
+    {
+
+    }
+    SSPBUF_REG = data;
+    #endif
```

Comment:
instead of duplicating a code we can have a one macro the defined according to MCU_TYPE 
```
#if MCU_TYPE == _AVR
#define SPI_WRITE_REG      SPDR
#define SPI_GET_TRANSFER_STATUS()      GET_BIT(SPSR, SPSR_SPIF)
#elif MCU_TYPE == _PIC
#define SPI_WRITE_REG      SSPBUF_REG
#define SPI_GET_TRANSFER_STATUS()     GET_BIT(SSPSTAT_REG, SSPSTAT_BF)
#else 
#error "Unkown mcu"
#endif 
```
and this is new implementation of the new function:
```
void SPI_Write(uint8_t data)
{
     SPI_WRITE_REG = data;
     while (SPI_GET_TRANSFER_STATUS() == 0);
```

Reply:
also we should consider handling this busy loop instead of stuck on it forever, this can be done by using overflow concept inside loop